### PR TITLE
Add zlib external coder

### PR DESCRIPTION
**Pull request by shubhamchandak94**
Add zlib external coder + interface for reading/writing binary files for tests

### CODE REVIEW

File: core/data_stream.py [hunk 1]
```diff
@@ -236,12 +236,26 @@ def write_symbol(self, s):
 
 
 class Uint8FileDataStream(FileDataStream):
-    """reads Uint8 numbers written to a file
```
Comment (kedartatwawadi) [inline core/data_stream.py:239]: Please add a short test ensuring that this is lossless

Comment (shubhamchandak94) [reply core/data_stream.py:239]: will do, thanks

File: compressors/external_compressors.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    create_random_binary_file,
+    try_file_lossless_compression,
+    try_lossless_compression,
+)
+import zlib
+
+
+class ZlibExternalEncoder(DataEncoder):
+    def __init__(self, level=6):
+        self.level = level
```
[the remark this answers is not in the document]
Comment (shubhamchandak94) [reply compressors/external_compressors.py:22]: Yes, we should do that. Created an issue to remember https://github.com/kedartatwawadi/stanford_compression_library/issues/47#issue-1266864041

File: compressors/external_compressors.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        # might be inefficient to convert to BitArray since it will be later be
+        # converted back to bytes when writing to file
+        # also, should we worry about endianness?
+
+        # at start write the compressed size because zlib decoder expects to get complete blocks
+        # and cannot determine when we want to end
+        compressed_bitarray = BitArray(
+            uint_to_bitarray(len(compressed_bytes) * 8, bit_width=self.block_size_num_bits)
+        )
+        compressed_bitarray.frombytes(compressed_bytes)
```
Comment (kedartatwawadi) [inline compressors/external_compressors.py:51]: why are we setting compressed_bitarray twice?

Comment (shubhamchandak94) [reply compressors/external_compressors.py:51]: First one just puts the length in there. Second one actual appends the data to it.

File: compressors/external_compressors.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        # state stays alive across blocks so we can benefit
+        self.zlib_context = zlib.compressobj(level=self.level)
+        self.block_size_num_bits = 32  # num bits used to encode the block size at start
+
+    def reset(self):
+        # start new zlib context
+        self.zlib_context = zlib.compressobj(level=self.level)
+
+    def encode_block(self, data_block: DataBlock):
+        raw_bytes = bytes(data_block.data_list)
```
[the remark this answers is not in the document]
Comment (shubhamchandak94) [reply compressors/external_compressors.py:32]: The DataBlock is actually a list, so I did this (even for a string I think we do a list of chars).

File: compressors/external_compressors.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+from utils.bitarray_utils import BitArray, bitarray_to_uint, uint_to_bitarray
+from utils.test_utils import (
+    create_random_binary_file,
+    try_file_lossless_compression,
+    try_lossless_compression,
+)
+import zlib
+
+
+class ZlibExternalEncoder(DataEncoder):
```
Comment (kedartatwawadi) [inline compressors/external_compressors.py:20]: Out of curiosity what does Zlib use under the hood? Maybe add a link etc. to Zlib documentation here and in the decoder

Comment (shubhamchandak94) [reply compressors/external_compressors.py:28]: added comment in revision

File: compressors/external_compressors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shubhamchandak94) [reply compressors/external_compressors.py:40]: discussed in call

File: core/data_stream.py (hunk 1 above)
Comment (shubhamchandak94) [reply core/data_stream.py:239]: done in revision

File: compressors/external_compressors.py (hunk 5 above)
Comment (kedartatwawadi) [reply compressors/external_compressors.py:28]: thanks!

**Pull request merged**